designs/plugin-resolution.md

```diff
@@ -0,0 +1,88 @@
+   configured during CLI creation for the above project version (`cli.New(cli.WithDefaultPlugins(...))`).
+    * In case both a project configuration file and the flag are found, fail if they are different.
+3. In case any of the plugin keys is not fully qualified (full name and version), check if there is only one plugin that
+   fits that unqualified plugin key.
+4. Resolve the plugins with the list of qualified keys above.
+5. Verify the plugins support the project version.
+   
+
+### Flaws
+
+* Default plugins need to be provided per project version
+* Project version can't be resolved (e.g., when a single project version is supported by the provided plugins).
+* Command calls are not allowed to override plugin keys.
+
+## Proposal
+
+The following plugin resolution algorithm solves the flaws described above:
+
+1. [All commands but Init] Obtain the project version and plugin keys from the project configuration if available.
+     * If they are available, plugin keys will be fully qualified.
+2. [Init command only] Obtain the project version from `--project-version` flag.
+     * The flag is optional, we will try to resolve project version later in case it is not provided.
+3. Obtain the optional plugin keys from `--plugins` flag.
+     * [Init command only] If not present, use default plugins.
+     * [All commands but Init] If not present, use plugin keys from configuration file (step 1).
+4. Qualify the unqualified keys from step 3.
+     * If we know the project version, use this info to filter the available plugins so that we can be more accurate.
+5. Resolve the plugins with the list of qualified keys above.
+6. [Init command only] Resolve the project version in case it wasn't provided and there is only one project version
+   supported by all the resolved plugins.
+   
+### Example
+
+Based on the following main.go file:
+```go
+package main
+
+import (
+   "log"
+
+   "sigs.k8s.io/kubebuilder/v2/pkg/cli"
+   "sigs.k8s.io/kubebuilder/v2/pkg/model/config"
+   pluginv2 "sigs.k8s.io/kubebuilder/v2/pkg/plugins/golang/v2"
+   pluginv3 "sigs.k8s.io/kubebuilder/v2/pkg/plugins/golang/v3"
+)
+
+func main() {
+   c, err := cli.New(
+      cli.WithCommandName("kubebuilder"),
+      cli.WithVersion(versionString()),
+      cli.WithPlugins(
+         &pluginv2.Plugin{},
+         &pluginv3.Plugin{},
+      ),
+      cli.WithDefaultPlugins(&pluginv3.Plugin{}),
+      cli.WithCompletion,
+   )
+   if err != nil {
+      log.Fatal(err)
+   }
+   if err := c.Run(); err != nil {
+      log.Fatal(err)
+   }
+}
+```
+
+As the `go.kubebuilder.io/v3` only supports project version "3", the project version can be resolved and doesn't
+need to be specified in the flags.
+```
+kubebuilder init --plugins=go.kubebuilder.io/v3
+```
+
+```
+kuebebuilder create api --plugins=declarative.kubebuilder.io/v1 ...
+```
+For this command, the declarative plugin will be used instead of the base plugin,
+but following command calls won't use it unless provided again.
```

Comment:
Let's imagine the scenario:

I scaffolded the project with SDK with `cli.WithDefaultPlugins(&golangv3.Plugin{}, &manifestsv1.Plugin{}, &scorecardv1.Plugin{}),`. Then, using kubebuilder I will run the create-api for this project. 

So, in POV it should work and KB CLI should raise the warning that the plugins  &manifestsv1.Plugin{}, &scorecardv1.Plugin{}) were not found and because of this not used. 

Could we add this example/scenario to the proposal as well?

Reply:
I dont think it should work. I think it should error out saying that manifest could not be resolved.